File: common/src/main/java/com/pranav/common/util/ZipUtil.java
package com.pranav.common.util;

import android.content.Context;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

public class ZipUtil {

    private static int BUFFER_SIZE = 1024 * 10;

    public static void unzipFromAssets(Context context, String zipFile, String destination) {
        try {
            var stream = context.getAssets().open(zipFile);
            unzip(stream, destination);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static void unzip(InputStream stream, String destination) {
        dirChecker(destination, "");
        byte[] buffer = new byte[BUFFER_SIZE];
        try {
            var zin = new ZipInputStream(stream);
            ZipEntry ze = null;
            while ((ze = zin.getNextEntry()) != null) {
                if (ze.isDirectory()) {
                    dirChecker(destination, ze.getName());
                } else {
                    var f = new File(destination, ze.getName());
                    if (!f.toPath().normalize().startsWith(destination))
                        throw new SecurityException(
                                "Potentially harmful files detected inside zip");
                    if (!f.exists()) {
                        if (!f.createNewFile()) {
                            continue;
                        }
                        var fout = new FileOutputStream(f);
                        int count;
                        while ((count = zin.read(buffer)) != -1) {
                            fout.write(buffer, 0, count);
                        }
                        zin.closeEntry();
                        fout.close();
                    }
                }
            }
            zin.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static void dirChecker(String destination, String dir) {
        var f = new File(destination, dir);

        if (!f.isDirectory()) {
            f.mkdirs();
        }
    }
}

File: common/src/main/java/org/cosmic/ide/common/util/ZipUtil.kt
package org.cosmic.ide.common.util

import android.content.Context

import java.io.IOException
import java.io.InputStream
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.io.File

object ZipUtil {

    const val BUFFER_SIZE = 2048 * 10

    @JvmStatic
    fun unzipFromAssets(context: Context, zipFile: String, destination: String) {
        try {
            val stream = context.assets.open(zipFile)
            unzip(stream, File(destination))
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun unzip(stream: InputStream, targetDir: File) {
        ZipInputStream(stream).use { zipIn ->
            var ze: ZipEntry? = zipIn.nextEntry
            while (ze != null) {
              val resolved = targetDir.resolve(ze.name).normalize()
              if (!resolved.startsWith(targetDir)) {
                    // see: https://snyk.io/research/zip-slip-vulnerability
                    throw SecurityException("Entry with an illegal path: " + ze.name)
                }
                if (ze.isDirectory) {
                    resolved.mkdirs()
                } else {
                    resolved.parentFile.mkdirs()
                    resolved.outputStream().use { output ->
                        zipIn.copyTo(output)
                    }
                }

                ze = zipIn.nextEntry
            }
        }

        stream.close()
    }
}
